protocol/interop-access-list.md

```diff
@@ -0,0 +1,103 @@
+    assembly {
+        let startGas := gas()
+        result := sload(_slot)
+        let endGas := gas()
+        isWarm := iszero(gt(sub(startGas, endGas), WARM_READ_COST))
+    }
+ }
+
+function validateMessage(Identifier calldata _id, bytes32 _msgHash) external {
+    bytes32 checksum = calculateChecksum(_id, _msgHash);
+
+    (bool _isSlotWarm,) = _isWarm(checksum);
+
+    if (!_isSlotWarm) revert NotWarm();
+
+    emit ExecutingMessage(_msgHash, _id);
+}
+```
+
+### Diagram
+
+Example using the `L2toL2CDM` with access-list in the transaction. The `L2toL2CDM` is not needed, it is being used for diagram clarity, any contract can call the `validateMessage` function.
+
+> Off-chain parts were skipped for simplification. The actual storage keys include more data not only the checksum. More info can be found on the specs doc.
+
+```mermaid
+sequenceDiagram
+    Relayer->>L2toL2CDM: tx with access-list: <br>{ "address": address(CrossL2Inbox), <br>"storageKeys": [checksum(Identifier, msgHash)] }
+    L2toL2CDM->>+CrossL2Inbox: validateMessage(Identifier, msgHash)
+    CrossL2Inbox-->>CrossL2Inbox: isWarm( checksum(Identifier, msgHash) )
+    Note over CrossL2Inbox: Gas cost < WARM_READ_COST
+    CrossL2Inbox-->>CrossL2Inbox: emit ExecutingMessage(msgHash, Identifier)
+    CrossL2Inbox->>L2toL2CDM: success
+```
+
+Example using the `L2toL2CDM` without access-list in the transaction. The `L2toL2CDM` is not needed, it is being used for diagram clarity, any contract can call the `validateMessage` function.
+
+```mermaid
+sequenceDiagram
+    Relayer->>L2toL2CDM: tx without access-list
+    L2toL2CDM->>+CrossL2Inbox: validateMessage(Identifier, msgHash)
+    CrossL2Inbox-->>CrossL2Inbox: isWarm( checksum(Identifier, msgHash) )
+    Note over CrossL2Inbox: Gas cost > WARM_READ_COST
+    CrossL2Inbox-->>L2toL2CDM: revert NotWarm()
+```
+
+### Alternatives considered
+
+Alternative solutions that kept the current dynamic declaration approach were explored. The most prominent one involved adding locking and unlocking functionality to the `CrossL2Inbox`. If the transaction referenced an invalid message, then that transaction would be included at the top of the next block when the `CrossL2Inbox` was locked. Because it was locked, the invalid transaction would revert. If it referenced an valid transaction, then it would be included when the `CrossL2Inbox` was unlocked. A special transaction took care of locking and unlocking the `CrossL2Inbox`.
+
+### Security considerations
+
+- Future changes in gas cost can affect this solution
+
+- Every message requires at most 3 entries in the access list and at least 2, therefore it will warms at most 3 and at least 2 storage slots, of which we only care about one. It's important that the checksum of the `Identifier` and `messageHash` data parameters provided to the `validateMessage` function doesn't result in one of these slots. We should ensure this scenario is statistically infeasible.
```

Comment:
```suggestion
- Every message requires at most 3 entries in the access list and at least 2, therefore it will warms at most 3 and at least 2 storage slots, of which we only care about one. It's important that the checksum of the `Identifier` and `messageHash` data parameters provided to the `validateMessage` function doesn't result in a slot warmed for any other reason than a prior non-reverting `validateMessage`. We should ensure there is no accidental storage collision that warms up an unverified checksum.
```